### PR TITLE
Add post mortem for Sept. 3, 2020

### PR DESCRIPTION
**Pull request by dmarcoux**

### CODE REVIEW

**Comment by hennevogel** [comment]
Deploy preview for *openbuildservice* ready!

Built with commit 7a8a34c000c94c4c24d7a2935c60cdd6c457745b

https://deploy-preview-268--openbuildservice.netlify.app

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+## Summary
+
+After a deployment, the search process running on our server crashed after we've
+attempted to fix an issue in our this process is tracked and monitored. The
+search was unavailable for 10 minutes.
+
+## Impact
+
+Nobody could search.
```
Comment (eduardoj) [inline _posts/deployments/2020-09-03-post-mortem.md:22]: I will also add that every part of the application that required a callback to update the Sphinx indices, returned a 500 error. That gave the false impresion to the users that something was really wrong with OBS, when only the update to the Sphinx indices was affected.

And until the reindex procces was finished, searching for some terms didn't return all the results expected.

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+### What Went Well
+
+We were quickly aware of the issue with the help of our errors tracker.
+
+### What Went Wrong
+
+We couldn't restart the systemd service after the searchd process crashed since
+searchd left out files which we had to remove.
+
```
Comment (eduardoj) [inline _posts/deployments/2020-09-03-post-mortem.md:62]: These left out files we had to remove, were created by starting up the daemon with a direct call to the Sphinx daemon, as root. We should have started the daemon with the `systemctl start obs-sphinx.service` command.

**Review by eduardoj — CHANGES_REQUESTED** [review]
In my honest opinion, there were some things missing, that after being added would help to clarify everything what happened.

File: _posts/deployments/2020-09-03-post-mortem.md (hunk 1 above)
Comment (dmarcoux) [reply _posts/deployments/2020-09-03-post-mortem.md:22]: True, I will add this.

File: _posts/deployments/2020-09-03-post-mortem.md (hunk 2 above)
Comment (dmarcoux) [reply _posts/deployments/2020-09-03-post-mortem.md:62]: In the heat of the moment, I didn't remember exactly why it was the case. I will explain this better.

**Comment by dmarcoux** [comment]
@eduardoj Please have another look.

**Comment by eduardoj** [comment]
> @eduardoj Please have another look.

Nice, thanks!

**Review by eduardoj — APPROVED** [review]

**Comment by dmarcoux** [comment]
I pushed again to fix a typo I found.

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+After a deployment, the search process running on our server crashed after we've
+attempted to fix an issue in our this process is tracked and monitored. The
+search was unavailable for 10 minutes. Errors also happened whenever updating
+projects and packages due to a tight integration with the search engine.
+
+## Impact
+
+Nobody could search. Any code updating the search indices through callbacks
+caused errors (500) while the search was down. Project and package pages were
```
Comment (hennevogel) [inline _posts/deployments/2020-09-03-post-mortem.md:24]: We know how many people where affected by this?

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+affected by this. Once it was back, we had partial search results while we
+reindexed the search.
+
+## Root Causes
+
+After a deployment on September 3, 2020 at 13:57 (CEST), the systemd service
+for Sphinx (our search engine) couldn't restart. The searchd process was still
+running, but somehow, the PID file of the systemd service didn't exist. Instead
+of leaving this in this problematic state, we fixed it, but also caused issues
+while doing so.
```
Comment (hennevogel) [inline _posts/deployments/2020-09-03-post-mortem.md:34]: So the root cause is `sphinx` crashing during the deployment, leaving behind unclean state.

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+## Trigger
+
+We were able to restart the systemd service after creating the missing PID file
+referring to the searchd process. However, at 14:31 (CEST), the systemd service
+stopped when the searchd process crashed and couldn't be restarted.
```
Comment (hennevogel) [inline _posts/deployments/2020-09-03-post-mortem.md:40]: This is not the trigger. What made sphinx crash?

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+10 minutes, outside of our maintenance window.  We want to give you some insight
+into what happened.
+
+- **Date**: 2020-09-03
+- **Authors**: Dany
+
+## Summary
+
+After a deployment, the search process running on our server crashed after we've
+attempted to fix an issue in our this process is tracked and monitored. The
```
[the remark this answers is not in the document]
Comment (dmarcoux) [reply _posts/deployments/2020-09-03-post-mortem.md:17]: Oh right, `in our this` should be `in how this process`

File: _posts/deployments/2020-09-03-post-mortem.md (hunk 3 above)
Comment (dmarcoux) [reply _posts/deployments/2020-09-03-post-mortem.md:24]: I don't remember the exact number, I think we had about 200 errors in Errbit.

**Comment by hennevogel** [comment]
Timeline:

- 2020-08-27 06:09:29 We rebuild the search indexes as root, which started the searchd as root
- everything was happily working
- 2020-09-03 11:57:17 deployment
- 2020-09-03 11:59:27 noticed that the search service failed to start
- 2020-09-03 12:00:34 noticed a dangling searchd process
- 2020-09-03 12:25:31 tried to fiddle with pid file
- 2020-09-03 12:34:55 noticed wrong permissions on search indexes
- 2020-09-03 12:40:41 we removed the index files, this didn't help
- 2020-09-03 12:47:11 started the searchd service as root, this worked
- 2020-09-03 12:57:27 stoppped the searchd service as root gracefully
- 2020-09-03 12:57:39 started the searchd service with systemd
- 2020-09-03 13:06:53 started reindexing

**Comment by dmarcoux** [comment]
I got some help from @hennevogel to better investigate and understand what happened. Please have another look.

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+- **Date**: 2020-09-03
+- **Authors**: Dany and Henne
+
+## Summary
+
+After a deployment, the searchd process running on our server crashed after
+we've attempted to fix an issue in how this process is tracked and monitored.
+The search was unavailable for 13 minutes. Errors also happened whenever
+updating projects and packages due to a tight integration with the search
+engine.
```
Comment (hennevogel) [inline _posts/deployments/2020-09-03-post-mortem.md:20]:
```suggestion
On 2020-08-27 we fixed an issue with our search indexes that required us to rebuild them. We did so and everything was running fine. Then on 2020-09-03 we noticed after a deployment that the search service did not start successfully. During our attempt to mitigate this, we destroyed the search indexes again which lead to this service degradation. The search was unavailable for 13 minutes. User visible errors also happened whenever projects and packages where created/updated due to a tight integration of those object with the search engine.
```

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+## Impact
+
+Nobody could search. Any code updating the search indices through callbacks
+caused errors (500) while the search was down. Project and package pages were
+affected by this. Once it was back, we had partial search results while we
+reindexed the search.
+
+## Root Causes
+
+We rebuilt the search indices as root on August 27th, 2020 at 8:09 (CEST).
```
Comment (hennevogel) [inline _posts/deployments/2020-09-03-post-mortem.md:31]:
```suggestion
To fix issue [#9436](https://github.com/openSUSE/open-build-service/issues/9436) we rebuilt the search indices August 27th, 2020 at 8:09 (CEST). Unfortunately we ran this command as the user `root`.
```

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+## Trigger
+
+After deploying on September 3, 2020 at 13:57 (CEST), we noticed that the
+systemd service for the search couldn't restart. We found out that the search
+indices were created as root. This created files as root and the searchd process
+couldn't be managed by its systemd service. We created a PID file for the
+systemd service in an attempt to have the systemd service manage the process,
+but it made the searchd process crash.
```
Comment (hennevogel) [inline _posts/deployments/2020-09-03-post-mortem.md:40]:
```suggestion
After deploying on September 3, 2020 at 13:57 (CEST), we noticed that the searchd systemd unit couldn't restart. While trying to mitigate this we removed search indexes (that also had wrong ownership) which caused the searchd service to fail. 
```

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+## Resolution
+
```
Comment (hennevogel) [inline _posts/deployments/2020-09-03-post-mortem.md:43]:
```suggestion
After a while we noticed that there was a searchd process running that was not controlled by the unit.
```

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+### What Went Well
+
+We were quickly aware of the issue with the help of our errors tracker.
+
+### What Went Wrong
+
+We weren't notified about the failure of systemd service for the search. This
+prevented us from knowing that we made a mistake by indexing the search as root.
+We also had to rebuild the search indices.
```
Comment (hennevogel) [inline _posts/deployments/2020-09-03-post-mortem.md:70]:
```suggestion
- We don't guard ourselves against running "standard" Ruby on Rails commands in production as the wrong user
- Our monitoring for the search service is not sophisticated enough
- The integration of the ActiveRecord callbacks and searchd is too fragile
- We removed search indices without thinking about the consequences
```

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+### Where We Got Lucky
+
+The issue was quite severe, but not the whole application went down with it.
+
+## Timeline (times in CEST)
+
+- **2020-08-27 08:09**: We rebuild the search indices as root, which created files for search indices as root
```
Comment (hennevogel) [inline _posts/deployments/2020-09-03-post-mortem.md:78]:
```suggestion
- **2020-08-27 08:09**: We rebuild the search indices as root, which started searchd as root and created search indices as that user
```

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+- Everything was working
```
Comment (hennevogel) [inline _posts/deployments/2020-09-03-post-mortem.md:79]:
```suggestion
- Everything was seemingly working fine. However, the `searchd` systemd unit was failing which we don't monitor. We only monitor for the existence of a process named `searchd`
```

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+- **2020-09-03 13:57**: Deployment
+- **13:59**: We notice that the systemd service of the search failed to start
+- **14:00**: We notice a dangling searchd process
+- **14:25**: We create a PID file in an attempt to have the systemd service
+    manage the dangling process
+- **14:34**: We notice wrong file permissions on the search indices. The searchd
+    process crashes. **The search is down**
```
Comment (hennevogel) [inline _posts/deployments/2020-09-03-post-mortem.md:86]:
```suggestion
- **14:34**: We notice wrong file permissions on the search indices and a message about this in the searchd logs
```

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+- **14:40**: We remove the files for the search indices
```
Comment (hennevogel) [inline _posts/deployments/2020-09-03-post-mortem.md:87]:
```suggestion
- **14:40**: In an desperate attempt, we remove the indices. **The search is down**
```

File: _posts/deployments/2020-09-03-post-mortem.md [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+- **14:47**: We start the searchd process as root again, **the search is no longer down**
+- **14:57**: We stop the searchd process gracefully
```
Comment (hennevogel) [inline _posts/deployments/2020-09-03-post-mortem.md:89]:
```suggestion
- **14:57**: We stop the searchd process gracefully, which briefly degrades the service again.
```

**Review by hennevogel — CHANGES_REQUESTED** [review]

**Review by hennevogel — APPROVED** [review]

**Pull request merged**